Source/DafnyRuntime/DafnyRuntime.cs

```diff
@@ -1189,15 +1189,23 @@ internal class ConcatSequence<T> : Sequence<T> {
       // Traverse the tree formed by all descendants which are ConcatSequences
       var ansBuilder = ImmutableArray.CreateBuilder<T>();
       var toVisit = new Stack<ISequence<T>>();
-      toVisit.Push(right);
-      toVisit.Push(left);
+      var (leftBuffer, rightBuffer) = (left, right);
+      if (left == null || right == null) {
+        return elmts;
+      }
+      toVisit.Push(rightBuffer);
+      toVisit.Push(leftBuffer);
 
       while (toVisit.Count != 0) {
         var seq = toVisit.Pop();
-        var cs = seq as ConcatSequence<T>;
-        if (cs != null && cs.elmts.IsDefault) {
-          toVisit.Push(cs.right);
-          toVisit.Push(cs.left);
+        if (seq is ConcatSequence<T> { elmts: { IsDefault: true } } cs) {
+          (leftBuffer, rightBuffer) = (cs.left, cs.right);
+          if (cs.left == null || cs.right == null) {
+            toVisit.Push(cs);
```

Comment:
Doesn't this cause an infinite loop?

Reply:
Not unless someone tries really hard. `left` and `right` are non-null when the sequence is fresh, otherwise the `Count` call would fail. The only place they are altered thereafter is `ImmutableElements`, which, assuming `ComputeElements` works, will always undefault `elmts`. In summary, there should be at most one retry.

Reply:
I don't know .NET's memory model as well as Java's - can we definitively say that if one thread sets `elmts` and then unsets `left` and `right`, then any other thread will see some prefix of those state changes in that order, even when there is no synchronization between them?

Reply:
I think I'm missing something easy :/ Assume a stack with a single element `cs` such that `IsDefault` is `true` and `cs.left` is `null`.  Won't the code pop `cs`, take the `seq is ConcatSequence<T>` branch, take the `cs.left == null` branch, push `cs` back into `toVisit`, and on the next iteration `pop` the same `cs`, thus getting into a loop?

Looking more at the code it looks like it's not really possible to have `IsDefault` set to `true` while `left` is `null`?  But in that case I don't understand when this branch will be taken (is it related to the concurrency problem?)

Reply:
> A write of a volatile field is called a volatile write. A volatile write has “release semantics”; that is, it is guaranteed to happen after any memory references prior to the write instruction in the instruction sequence.
The write is happening in ImmutableElements, so we should be good.

[ECMA-334](https://www.ecma-international.org/wp-content/uploads/ECMA-334_5th_edition_december_2017.pdf) _15.5.4 Volatile fields_

That's exactly what we need for `ImmutableElements`.

Reply:
> Looking more at the code it looks like it's not really possible to have IsDefault set to true while left is null? But in that case I don't understand when this branch will be taken (is it related to the concurrency problem?)

Yes that is to avoid concurrency issues. The idea is that the condition in L1201 becomes `false` by the time we hit L1203. We know that this will happen up to once, so we retry expecting to end up in L1210.

Reply:
No, because we want to keep going. There might be more on the stack.

Reply:
Alright, i understand now (@fabiomadge explained ^^).  Let's add a comment in the code that points out the mutation.